eng/common/performance/performance-setup.sh

```diff
@@ -107,6 +112,7 @@ while (($# > 0)); do
       echo "  --kind <value>                 Related to csproj. The kind of benchmarks that should be run. Defaults to micro"
       echo "  --runcategories <value>        Related to csproj. Categories of benchmarks to run. Defaults to \"coreclr corefx\""
       echo "  --internal                     If the benchmarks are running as an official job."
+      echo "  --monodotnet                   Pass the path to the mono dotnet for mono performance testing."
```

Comment:
I don't think we should do this. It'd be better to just use self-contained build

/cc @directhex @akoeplinger

Reply:
Can you explain this a little more? Why is using the patched mono dotnet not a good path forward? This is similar to what we do for the coreclr based runtime tests.

Also, the way that we currently run performance tests relies on having a version of dotnet to build and run the Benchmark DotNet tests, and as far as I know we cannot do a self-contained publish of the performance tests. (@billwert for confirmation).

Reply:
> as far as I know we cannot do a self-contained publish of the performance tests. 

I have been doing that for months on my infrastructure so at least in theory it's possible. On the other hand it's not easy and useful for most cases because BDN by default triggers the compilation for the benchmarks and runs each of them out of process.

For non-AOT benchmarks I ended up using/patching `testhost` that is produced as output of the dotnet/runtime repository. It it guaranteed to be self-consistent unlike the combination of SDK version determined by global.json and the runtime built from the repository.

Reply:
> Why is using the patched mono dotnet not a good path

This is not set up we'll ever officially support and can break anything as it was a temporary hack. It also won't support the official AOT mode and possibly other configs. So if you are testing this hack then you are testing config which no customer will use

Reply:
For some context, we run the performance testing for the coreclr runtime in two ways. The first is on the pacakges from the installer repo. This is to make sure that we are running the performance tests in a way that the customer will actually use the product. We also run the tests here in the runtime repo by building the product and then making a corerun directory that contains all of our built binaries. This, while not the way our customers use or obtain the product, allows us to get much more granular performance data collected as we don't have to wait for code to flow all the way from here to the installer repo.

This work was designed to get mono to parity with what we are doing here in the runtime repo. So building the product from head and then making up a version of dotnet that contains the built binaries and running the tests. If using the dotnet-mono patch feature is not the right way to do this, please direct me to the canonical way to do this. I was just operating off of the steps given to me by @SamMonoRT. 

@filipnavara:

1. When you say you have been doing this for months, do you mean the tests that live in the dotnet/performance repo? If so I would love to get instructions on how you are doing that.
2. I would also be interested in how you are doing the AOT testing as well. This is something that we would like to be able to support y'all on, so getting an idea of how y'all are building and running these tests, even if we have to take a different path would be useful.

Reply:
> This work was designed to get mono to parity with what we are doing here in the runtime repo. So building the product from head and then making up a version of dotnet that contains the built binaries and running the tests. If using the dotnet-mono patch feature is not the right way to do this, please direct me to the canonical way to do this. I was just operating off of the steps given to me by @SamMonoRT.

I am doing exactly that for my JIT runs. The difference is that I start off with the `testhost` that is built from this repository instead of the SDK. Then I just do
```
      cp -r runtime/artifacts/bin/testhost/${TESTHOST_NAME}/ mono-testhost/
      cp -r runtime/artifacts/bin/mono/${RUNTIME_CONFIGURATION}/* mono-testhost/shared/Microsoft.NETCore.App/5.0.0/
```
which produces a directory with Mono runtime, libraries and a `dotnet` executable that can be passed to BDN. Now I believe that such variation of testhost may already exist in some build configuration directly since that's exactly the same environment that is used for running unit tests on each PR.

> 1. When you say you have been doing this for months, do you mean the tests that live in the dotnet/performance repo? If so I would love to get instructions on how you are doing that.

Yes. I run daily scheduled builds for the Mono JIT using the Azure pipeline hosted at https://github.com/filipnavara/mono-performance/blame/master/azure-pipelines.yml. Maybe something out of it could be useful, maybe not. 

> 2. I would also be interested in how you are doing the AOT testing as well. This is something that we would like to be able to support y'all on, so getting an idea of how y'all are building and running these tests, even if we have to take a different path would be useful.

I'd love to give you the instructions but at the moment it's not an automated process for me (unlike the JIT runs). The basic steps I use are the following:
- Get a patched dotnet SDK with Mono runtime (not essential)
- Build a self-contained version of the benchmark using `dotnet publish -f netcoreapp5.0 -c Release -r osx-x64`
- Patch the runtime in the output directory by replacing libcoreclr.dylib / System.Private.CoreLib.{dll,pdb} (unless already done by the patched SDK)
- For the AOTing I used this `MONO_PATH=../../../artifacts/bin/MicroBenchmarks/Release/netcoreapp5.0/osx-x64/ ../../../../mono/mono/mini/mono-sgen --aot=llvm --assembly-loader=strict ../../../artifacts/bin/MicroBenchmarks/Release/netcoreapp5.0/osx-x64/System.Private.CoreLib.dll` applied for all the binaries in the publish output. Now this has changed by the move from `mono/mono` repository to `dotnet/runtime` but it's possible to do the same thing through the `dotnet` executable with the right parameters (@EgorBo, am I right?).
- Finally I run the output using `MONO_PATH=../../../artifacts/bin/MicroBenchmarks/Release/netcoreapp5.0/osx-x64/ ../../../../mono/mono/mini/mono-sgen --assembly-loader=strict ../../../artifacts/bin/MicroBenchmarks/Release/netcoreapp5.0/osx-x64/MicroBenchmarks.dll --anyCategories coreclr --iterationCount 1 --warmupCount 0 --invocationCount 1 --unrollFactor 1 --strategy ColdStart --stopOnFirstError true -i`. Again, `mono-sgen` can be replaced by `dotnet` here. The only important parameter is the `-i` that forces the in-process mode. The number of iterations is reduced intentionally since there's no JIT overhead.

Reply:
> This, while not the way our customers use or obtain the product, allows us to get much more granular performance data collected as we don't have to wait for code to flow all the way from here to the installer repo.

We have something similar (which is not this hack) but allow to reference in tree runtime packs. This could work with coreclr as well if you would like to unify the approaches.

/cc @steveisok

Reply:
The in-tree runtime pack option only works for mobile right now and would need a few tweaks to get that to work for desktop.  I do see that as a good replacement for a patched mono.  

I guess the question is should we make this PR depend on it or not?

Reply:
We should not block this PR till the in-tree runtime pack is fully working for desktop. That should be a future todo.

Reply:
So it seems like a good place to land here is to move the performance runs over to the testhost construction, as that seems to be what y'all have been doing in the past. We can then go ahead and get this checked in and running for Windows and Linux and for LLVM and non-LLVM backend. Then we can file an issue that tracks doing the work to move over to the in tree runtime packs as soon as we have good support for that on desktop builds. Does that sound like a good way forward to everyone here?

Reply:
Sounds good to me - @steveisok @marek-safar thoughts ?

Reply:
AOT won't work this way for sure and any issues with this setup won't be fixed. I'm fine to land this but we should start working on a supported way to run the tests as this can break anything.

Reply:
So I did some digging and this is the step that seems to setup the testhost for running functional tests on mono, [link](https://dev.azure.com/dnceng/public/_build/results?buildId=606124&view=logs&jobId=175a96bc-101d-5851-5ab5-f85186176673&j=c6f8dc49-92a1-5760-c098-ba97b8142bfb&t=5d5289b6-723b-5155-5d15-b590d280a14a). I would assume that since this is how the product is tested in CI that this should be a supported scenario, yes?

If this is the case I am just going to use the same command to setup the testhost and run the performance tests in this way for now.

If this is also unsupported does that mean that the functional CI tests are running in an unsupported manner or am I just missing something here?

Reply:
I think that's as close to supported desktop scenario as it is currently available (until desktop runtime packs are introduced).